File: src/main/java/net/clgd/ccemux/Runner.java
package net.clgd.ccemux;

import java.awt.SplashScreen;
import java.nio.file.Path;
import java.util.AbstractMap.SimpleEntry;
import java.util.stream.Collectors;
import java.util.HashMap;
import java.util.List;

import org.slf4j.Logger;
import org.squiddev.cctweaks.lua.lib.ApiRegister;

import net.clgd.ccemux.emulation.CCEmuX;
import net.clgd.ccemux.emulation.EmulatedComputer;
import net.clgd.ccemux.rendering.Renderer;
import net.clgd.ccemux.rendering.RenderingMethod;

public class Runner {
	public static void launch(Logger logger, Config config, List<Path> saveDirs, int count) {
		CCEmuX emu = new CCEmuX(logger, config);
		HashMap<EmulatedComputer, List<Renderer>> computers = new HashMap<>();

		for (int i = 0; i < count; i++) {
			EmulatedComputer ec = emu.createEmulatedComputer(-1, saveDirs.size() > 0 ? saveDirs.remove(0) : null);
			computers.put(ec, emu.conf.getRenderer().stream().map(r -> RenderingMethod.create(r, emu, ec))
					.collect(Collectors.toList()));
		}
		
		if (SplashScreen.getSplashScreen() != null) SplashScreen.getSplashScreen().close();
		
		computers.forEach((ec, r) -> r.forEach(r2 -> r2.setVisible(true)));
		
		emu.run();
	}

	public static void launchCCTweaks(Logger logger, Config config, List<Path> saveDirs, int count) {
		config.addListener(new Runnable() {
			@Override
			public void run() {
				config.entrySet().stream().map(e -> new SimpleEntry<>((String) e.getKey(), (String) e.getValue()))
						.filter(e -> e.getKey().startsWith("cctweaks"))
						.forEach(e -> System.setProperty(e.getKey(), e.getValue()));
			}
		});

		org.squiddev.patcher.Logger.instance = new org.squiddev.patcher.Logger() {
			@Override
			public void doDebug(String message) {
				logger.debug(message);
			}

			@Override
			public void doWarn(String message) {
				logger.warn(message);
			}

			@Override
			public void doError(String message, Throwable t) {
				logger.error(message, t);
			}
		};

		ApiRegister.init();

		launch(logger, config, saveDirs, count);
	}
}

File: src/main/kotlin/net/clgd/ccemux/Runner.kt
package net.clgd.ccemux

import net.clgd.ccemux.emulation.CCEmuX
import net.clgd.ccemux.emulation.EmulatedComputer
import net.clgd.ccemux.rendering.BlankRenderer
import net.clgd.ccemux.rendering.Renderer
import net.clgd.ccemux.rendering.RenderingMethod
import org.slf4j.Logger
import org.squiddev.cctweaks.lua.ConfigPropertyLoader
import org.squiddev.cctweaks.lua.lib.ApiRegister
import java.awt.SplashScreen
import java.nio.file.Path
import java.util.*

object Runner {
	@JvmStatic fun launchCCTweaks(logger: Logger, config: Config, saveDirs: MutableList<Path>, count: Int) {
		// Add config listener to sync CCTweaks config.
		// This allows changing it at runtime, should you ever wish to do that.
		// This needs to be done in this class as the config class should be loaded within the
		// wrapped class loader.
		config.addListener(Runnable {
			config.entries.forEach {
				p -> run {
					val key = p.key

					if (key != null && key is String && key.startsWith("cctweaks")) {
						val value = p.value

						if (value != null && value is String) {
							System.setProperty(key, value)
						}
					}
				}
			}

			ConfigPropertyLoader.init()
		})

		// Redirect CCTweaks' logger
		org.squiddev.patcher.Logger.instance = object : org.squiddev.patcher.Logger() {
			override fun doDebug(message: String) = logger.debug(message)

			override fun doWarn(message: String) = logger.warn(message)

			override fun doError(message: String, e: Throwable) = logger.error(message, e)
		}

		// Various setup tasks for CCTweaks
		ApiRegister.init()

		// And launch!
		launch(logger, config, saveDirs, count)
	}

	@JvmStatic fun launch(logger: Logger, config: Config, saveDirs: MutableList<Path>, count: Int) {
		val emu = CCEmuX(logger, config)

		val computers = HashMap<EmulatedComputer, List<Renderer>>()

		for (i in 0..(count-1)) {
			val it = if (saveDirs.isNotEmpty()) {
				emu.createEmulatedComputer(saveDirs.removeAt(0))
			} else {
				emu.createEmulatedComputer()
			}

			computers.put(it, emu.conf.getRenderer().map { r -> RenderingMethod.create(r, emu, it) ?: BlankRenderer() })
		}

		SplashScreen.getSplashScreen()?.close()
		computers.forEach { ec, r -> r.forEach { f -> f.setVisible(true) } }
		emu.run()
	}
}
